R/GeospatialFunctions.R

```diff
@@ -0,0 +1,607 @@
+      dplyr::select(-c(index, epsg))
+  }))
+  
+  return(sf)
+  
+}
+
+
+#' fetchATTAINS
+#' 
+#' Fetch ATTAINS features within a bounding box produced from a set of TADA spatial features.
+#' 
+#' @param data A dataframe developed using `TADA_DataRetrieval()` or `TADA_MakeSpatial()`.
+#' @param type The type of ATTAINS data you would like to fetch ("lines", "points", "polygons", "catchments"). All ATTAINS features are returned in WGS84 (crs = 4326).
+#' 
+#' @return spatial features that are within the spatial bounding box of water quality observations.
+#' 
+#' @seealso [TADA_MakeSpatial()]
+#' @seealso [TADA_DataRetrieval()]
+#' 
+#' @examples
+#' \dontrun{
+#'tada_data <- TADA_DataRetrieval(startDate = "1990-01-01",
+#'                                endDate = "1995-12-31",
+#'                                characteristicName = "pH",
+#'                                statecode = "NV",
+#'                                applyautoclean = TRUE)
+#'                                  
+#'nv_attains_lines <- fetchATTAINS(data = tada_data, type = "lines")
+#' }
+
+fetchATTAINS <- function(data, type = NULL) {
+  
+  if(is.null(data) | nrow(data) == 0){
+    stop("There is no data in your `data` object to use as a bounding box for selecting ATTAINS features.")
+  }
+  
+  # If data is already spatial, just make sure it is in the right CRS
+  # and add an index as the WQP observations' unique identifier...
+  if (!is.null(data) & inherits(data, "sf")) {
+    if(sf::st_crs(data)$epsg != 4326){
+      data <- data %>%
+        sf::st_transform(4326)
+    } else {
+      data <- data 
+    }
+  } else {
+    # ... Otherwise transform into a spatial object then do the same thing:
+    data <- data %>%
+      #convert dataframe to a spatial object
+      TADA_MakeSpatial(data = ., crs = 4326) 
+  }
+  
+  if(is.null(type)){
+    stop("Please select the type of ATTAINS data you would like to import ('catchments', 'lines', 'points', or 'polygons').")
+  }
+  
+  # select baseurl based on type of ATTAINS data you want to download:
+  if(type == "catchments"){
+    baseurl <- "https://gispub.epa.gov/arcgis/rest/services/OW/ATTAINS_Assessment/MapServer/3/query?"
+  }else if(type == "points"){
+    baseurl <-  "https://gispub.epa.gov/arcgis/rest/services/OW/ATTAINS_Assessment/MapServer/0/query?"
+  }else if(type == "lines"){
+    baseurl <-  "https://gispub.epa.gov/arcgis/rest/services/OW/ATTAINS_Assessment/MapServer/1/query?"
+  }else if(type == "polygons"){
+    baseurl <-  "https://gispub.epa.gov/arcgis/rest/services/OW/ATTAINS_Assessment/MapServer/2/query?"
+  }
+  
+  # starting at feature 1 (i.e., no offset):
+  offset <- 0 
+  # empty list to store all features in
+  all_features <- list()
+  
+  # EPSG we want our ATTAINS data to be in (always 4326 for this function)
+  epsg <- 4326
+  
+  # bounding box (with some minor wiggle) of user's WQP data
+  suppressMessages(suppressWarnings({
+    
+    bbox <- data %>% 
+      sf::st_buffer(0.001) %>% 
```

Comment:
I think the precision is not an issue... mostly because this is a precautionary step for pulling in ATTAINS data for an area extending beyond (0.001 DD) our WQP points, so this data ultimately gets (likely) majorly winnowed down as soon as we select only objects that intersect the catchments containing WQP data. But! This approach does introduce an unlikely/very rare risk of not capturing all the raw ATTAINS sf's (points, lines, polygons) associated with intersecting catchments. For example, there is a ATTAINS line outside the bbox, but overlays a catchment containing the WQP point (that happens to also be within a catchment with ATTAINS data to begin with). This is a real issue I would like to fix, but I don't know of an efficient solution to this problem....